src/serac/numerics/functional/element_restriction.hpp

```diff
@@ -0,0 +1,255 @@
+#pragma once
+
+#include <vector>
+
+#include "mfem.hpp"
+#include "axom/core.hpp"
+#include "geometry.hpp"
+
+inline bool isH1(const mfem::FiniteElementSpace& fes)
+{
+  return (fes.FEColl()->GetContType() == mfem::FiniteElementCollection::CONTINUOUS);
+}
+
+inline bool isHcurl(const mfem::FiniteElementSpace& fes)
+{
+  return (fes.FEColl()->GetContType() == mfem::FiniteElementCollection::TANGENTIAL);
+}
+
+inline bool isDG(const mfem::FiniteElementSpace& fes)
+{
+  return (fes.FEColl()->GetContType() == mfem::FiniteElementCollection::DISCONTINUOUS);
+}
+
+enum class FaceType
+{
+  BOUNDARY,
+  INTERIOR
+};
+
+/// a struct of metadata (index, sign, orientation) associated with a degree of freedom
+struct DoF {
+  // sam: I wanted to use a bitfield for this type, but a 10+ year-old GCC bug
+  // makes it practically impossible to assign to bitfields without warnings/errors
+  // with -Wconversion enabled, see: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=39170
+  // So, we resort to masks and bitshifting instead.
+
+  static constexpr uint64_t sign_mask        = 0x8000'0000'0000'0000;       ///< bits for sign field
```

Comment:
Can we use `std::vector<bool>` or `std::bitset`?  Is this required because it is used on the device?

Reply:
`std::vector<bool>` is notoriously weird and I don't think there's much of an advantage to `std::bitset` (in addition to the `__device__` compatibility issues). Were there specific concerns about this implementation?

Reply:
not really - just bitset syntax always takes me a bit longer to parse.